docs/features/pouch_with_upgrade.md

```diff
@@ -0,0 +1,102 @@
+
+## Design
+
+Before using the `upgrade` interface, we should figure out which parameter of container you want to update ? If you want to update a container's [Resources](../../apis/types/resources.go), I think you should use the `update` interface. If you just want to update a container's image and inherit the volumes and network from the old container, now you should use the `upgrade` interface.
+
+### About API
+
+Now, `upgrade` api request body is below, we only support you to specify the Image which used to create a new container and the new `Entrypoint` and `Cmd` that the new container used to run.
+
+```golang
+type ContainerUpgradeConfig struct {
+
+	// Execution commands and args
+	// Min Items: 1
+	Cmd []string `json:"Cmd"`
+
+	// The entrypoint for the container as a string or an array of strings.
+	// If the array consists of exactly one empty string (`[""]`) then the entry point is reset to system default.
+	//
+	Entrypoint []string `json:"Entrypoint"`
+
+	// image
+	// Required: true
+	Image string `json:"Image"`
+}
+```
+
+It is enough to update a container from a old image to a new one now. If anyone has new situation need to add more parameters into `ContainerUpgradeConfig`, we can create a new issue to discuss it.
+
+### About core logic
+
+#### Which cmd should be used to create the new container when upgrade
+
+There are four commands as mentioned below, when create a new container in upgrade, the order of using the command is:
+
+* We should first use CMD3;
+* We should use the CMD1, if CMD4 is not specified ;
+* We need to use CMD4, if CMD1 is also not specified ;
+* CMD2 should not be used.
+
+|cmd|named|
+|---|---|
+|Old cmd of `ContainerConfig`| CMD1|
+|Old cmd of old image| CMD2|
+|New cmd of `ContainerUpgradeConfig`|CMD3|
+|New cmd of new image|CMD4|
+
+#### Inherit the volumes and network from the old container when create new container
+
+* network: before create a new container, we should first stop the old container if it is running, then we can use the network config of the old container.
+
+* storage: new container inherits all the mounts information of the old one, so we can use the volumes used by old container, and we should also parse the volume information contained in new image.
+
+#### Prepare rootfs for the new container
+
+In [containerd](https://github.com/containerd/containerd), using a new snapshot to represent the rootfs of container when create a new container. So we should use the new image to create a new snapshot for the new container in `upgrade`. In case of snapshot name conflicted, we add a random suffix string to the container id as the name of the new snapshot.
+
+#### rollback of upgrade failure
+
+When occurred any error during upgrade, we should have the ability to rollback the upgrade and recover the old container just like it before.
```

Comment:
When any errors occur during `upgrade` procedure, pouchd will automatically rollback the whole `upgrade` operation and the old container will be recovered just like it was.